In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### Loading Dataset

In [6]:
df = pd.read_csv('../../../data/3-clean data/Data_Rework/waste_treatment_y_p_wt_32111-0011.csv')
# src\data\3-clean data\Data_Rework\waste_treatment_y_p_wt_32111-0011.csv

In [7]:
df.head()

Year              State                       Type of plant  \
0  2006  Baden-Württemberg                           Landfills   
1  2006  Baden-Württemberg      Thermal waste treatment plants   
2  2006  Baden-Württemberg                   combustion plants   
3  2006  Baden-Württemberg  Chemical-physical treatment plants   
4  2006  Baden-Württemberg               Soil treatment plants   

   Waste treatment plants\r\n(number)  \
0                                 416   
1                                  10   
2                                  24   
3                                  31   
4                                   3   

   Input to waste treatment plants\r\n(1000t)  \
0                                      5391.9   
1                                      1612.2   
2                                      1201.3   
3                                       267.7   
4                                        96.3   

   Wastes generated in the own local unit\r\n(1000t)  \
0                                               35.2   
1                                               93.9   
2                                              107.4   
3                                                3.2   
4                                                0.0   

   Waste delivered from the domestic territory\r\n(1000t)  \
0                                             5255.9        
1                                             1516.2        
2                                             1044.0        
3                                              250.5        
4                                               92.5        

   Waste delivered from abroad\r\n(1000t)  \
0                                   100.7   
1                                     2.1   
2                                    49.9   
3                                    14.1   
4                                     3.8   

   Waste treatment plants with output\r\n(number)  \
0                                               0   
1                                               0   
2                                               0   
3                                               0   
4                                               0   

   Output of waste treatment plants\r\n(1000t)  \
0                                        245.1   
1                                        437.5   
2                                         65.8   
3                                        123.4   
4                                         95.9   

   Wastes to be disposed of\r\n(1000t)  Wastes to be recovered\r\n(1000t)  \
0                                 64.2                              180.9   
1                                 11.2                              413.5   
2                                 12.2                               53.5   
3                                 10.4                              109.9   
4                                 37.6                               58.2   

   Waste to preparatory processes\r\n(1000t)  \
0                                        0.0   
1                                        0.0   
2                                        0.0   
3                                        0.0   
4                                        0.0   

   Distribution to other plants0 users or traders\r\n(1000t)  
0                                                0.0          
1                                               12.8          
2                                                0.0          
3                                                3.1          
4                                                0.0

In [68]:
from pandas_profiling import ProfileReport


### Data description 
Year - year 

State - place

Type of plant - Types landfils

Waste treatment plants -(number)

Input to waste treatment plants - I think 1000t refers to 1000 tons
(1000t)

Wastes generated in the own local unit
(1000t)

Waste delivered from the domestic territory
(1000t)

Waste delivered from abroad
(1000t)

Waste treatment plants with output
(number)

Output of waste treatment plants
(1000t)

Wastes to be disposed of
(1000t)

Wastes to be recovered
(1000t)

Waste to preparatory processes
(1000t)

Distribution to other plants0 users or traders
(1000t)

In [69]:
profile = ProfileReport(df)
profile

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

### Inference - 1 (Intial EDA)
There seems  to be no Null values in the dataset rather than null values there seems to be a surge of Zeros . We need to confirm whether 0 have been put to fill the na values , or whether there seems to be an issue with collection processs and thus requires imputation

Total rows = 3840 with 7.4% duplicates Need to further check it

**Year** - 2006 -2021 Data has been distributed equally among the years

**State** - 10 values Distribution of the Data
Mechlenburg-vovpommern - 25% data
Saarland = 18.8%
Hamburg - 12.5%
baden-wurttenberg - 12.1%
Rest have under 6.3% percent

**Types of wastes**
15 distinct values 
Landfills, Thermal waste treatment plants , Combution plants,Soil Treatment Plants, 
All the types have around 6.7 % of data so the sample data has equal distribution

**On the basis of Correlation**
We can observe that output of waste treatment plants, waste to be disposed of , waste to be recovered have really high postive correlation 

**Potential y variable**
At this point I feel Wastes generated in the own local unit can be used as a potential output variable

**Outlier detection** TODO!!!

##### Analysing 0 values

In [8]:
df_only_num = df.select_dtypes(include=[int,float])
df_only_num.columns

Index(['Year', 'Waste treatment plants\r\n(number)',
       'Input to waste treatment plants\r\n(1000t)',
       'Wastes generated in the own local unit\r\n(1000t)',
       'Waste delivered from the domestic territory\r\n(1000t)',
       'Waste delivered from abroad\r\n(1000t)',
       'Waste treatment plants with output\r\n(number)',
       'Output of waste treatment plants\r\n(1000t)',
       'Wastes to be disposed of\r\n(1000t)',
       'Wastes to be recovered\r\n(1000t)',
       'Waste to preparatory processes\r\n(1000t)',
       'Distribution to other plants0 users or traders\r\n(1000t)'],
      dtype='object')

In [9]:
df[df['Waste delivered from abroad\r\n(1000t)'] == 0]['Waste delivered from abroad\r\n(1000t)'].count() # 2326 columns have 0 values

# df_only_num[(df_only_num == 0).all(axis=1)]

2326

In [10]:
df[df['Wastes generated in the own local unit\r\n(1000t)'] == 0]

Year              State  \
4     2006  Baden-Württemberg   
6     2006  Baden-Württemberg   
7     2006  Baden-Württemberg   
10    2006  Baden-Württemberg   
12    2006  Baden-Württemberg   
...    ...                ...   
3832  2021          Thuringia   
3835  2021          Thuringia   
3836  2021          Thuringia   
3837  2021          Thuringia   
3838  2021          Thuringia   

                                        Type of plant  \
4                               Soil treatment plants   
6        Mechanical-biological waste treatment plants   
7     Dismantling facilities for end-of-life vehicles   
10                             Dismantling facilities   
12              natural substances produced in mining   
...                                               ...   
3832  Dismantling facilities for end-of-life vehicles   
3835                           Dismantling facilities   
3836                           Other treatment plants   
3837            natural substances produced in mining   
3838           Wastes at underground extraction sites   

      Waste treatment plants\r\n(number)  \
4                                      3   
6                                      3   
7                                    148   
10                                    23   
12                                     0   
...                                  ...   
3832                                  54   
3835                                   0   
3836                                   0   
3837                                   0   
3838                                   0   

      Input to waste treatment plants\r\n(1000t)  \
4                                           96.3   
6                                          217.3   
7                                           76.5   
10                                          19.3   
12                                           0.0   
...                                          ...   
3832                                         9.8   
3835                                         0.0   
3836                                         0.0   
3837                                         0.0   
3838                                         0.0   

      Wastes generated in the own local unit\r\n(1000t)  \
4                                                   0.0   
6                                                   0.0   
7                                                   0.0   
10                                                  0.0   
12                                                  0.0   
...                                                 ...   
3832                                                0.0   
3835                                                0.0   
3836                                                0.0   
3837                                                0.0   
3838                                                0.0   

      Waste delivered from the domestic territory\r\n(1000t)  \
4                                                  92.5        
6                                                 217.3        
7                                                  76.1        
10                                                 19.3        
12                                                  0.0        
...                                                 ...        
3832                                                9.7        
3835                                                0.0        
3836                                                0.0        
3837                                                0.0        
3838                                                0.0        

      Waste delivered from abroad\r\n(1000t)  \
4                                        3.8   
6                                        0.0   
7                                        0.3   
10                                       0.0   
12                                       0.0   
... 

In [11]:
for col in df_only_num.columns:
  print(col,'has' ,df_only_num[df_only_num[col] == 0][col].count(),'Percentage:',round(df_only_num[df_only_num[col] == 0][col].count()/len(df_only_num)*100,2),'% zeros present \n')

Year has 0 Percentage: 0.0 % zeros present 

Waste treatment plants
(number) has 1255 Percentage: 32.68 % zeros present 

Input to waste treatment plants
(1000t) has 1558 Percentage: 40.57 % zeros present 

Wastes generated in the own local unit
(1000t) has 2382 Percentage: 62.03 % zeros present 

Waste delivered from the domestic territory
(1000t) has 1593 Percentage: 41.48 % zeros present 

Waste delivered from abroad
(1000t) has 2326 Percentage: 60.57 % zeros present 

Waste treatment plants with output
(number) has 3825 Percentage: 99.61 % zeros present 

Output of waste treatment plants
(1000t) has 1862 Percentage: 48.49 % zeros present 

Wastes to be disposed of
(1000t) has 2178 Percentage: 56.72 % zeros present 

Wastes to be recovered
(1000t) has 1907 Percentage: 49.66 % zeros present 

Waste to preparatory processes
(1000t) has 3689 Percentage: 96.07 % zeros present 

Distribution to other plants0 users or traders
(1000t) has 2204 Percentage: 57.4 % zeros present 



In [12]:
df[df.duplicated() ==True]
#  Around 482 rows are duplicated 

Year                   State  \
163   2006  Mecklenburg-Vorpommern   
192   2006                Saarland   
267   2007       Baden-Württemberg   
403   2007  Mecklenburg-Vorpommern   
432   2007                Saarland   
...    ...                     ...   
3801  2021                Saarland   
3805  2021                Saarland   
3806  2021                Saarland   
3807  2021                Saarland   
3808  2021                Saarland   

                                     Type of plant  \
163         Wastes at underground extraction sites   
192          natural substances produced in mining   
267          natural substances produced in mining   
403         Wastes at underground extraction sites   
432          natural substances produced in mining   
...                                            ...   
3801  Mechanical-biological waste treatment plants   
3805                        Dismantling facilities   
3806                        Other treatment plants   
3807         natural substances produced in mining   
3808        Wastes at underground extraction sites   

      Waste treatment plants\r\n(number)  \
163                                    0   
192                                    2   
267                                    0   
403                                    0   
432                                    2   
...                                  ...   
3801                                   0   
3805                                   0   
3806                                   0   
3807                                   0   
3808                                   0   

      Input to waste treatment plants\r\n(1000t)  \
163                                          0.0   
192                                          0.0   
267                                          0.0   
403                                          0.0   
432                                          0.0   
...                                          ...   
3801                                         0.0   
3805                                         0.0   
3806                                         0.0   
3807                                         0.0   
3808                                         0.0   

      Wastes generated in the own local unit\r\n(1000t)  \
163                                                 0.0   
192                                                 0.0   
267                                                 0.0   
403                                                 0.0   
432                                                 0.0   
...                                                 ...   
3801                                                0.0   
3805                                                0.0   
3806                                                0.0   
3807                                                0.0   
3808                                                0.0   

      Waste delivered from the domestic territory\r\n(1000t)  \
163                                                 0.0        
192                                                 0.0        
267                                                 0.0        
403                                                 0.0        
432                                                 0.0        
...                                                 ...        
3801                                                0.0        
3805                                                0.0        
3806                                                0.0        
3807                                                0.0        
3808                                                0.0        

      Waste delivered from abroad\r\n(1000t)  \
163                                      0.0   
192                                      0.0   
267                                      0.0   
403                                      0.0   
432                         

In [13]:
### Finding out rows which have all the numerical columns as null values

# df[df['Wastes generated in the own local unit\r\n(1000t)'] == 0 & ('Input to waste treatment plants\r\n(1000t)' == 0) & ('Waste delivered from the domestic territory\r\n(1000t)' == 0) & ('Waste delivered from abroad\r\n(1000t)' == 0) & ('Waste treatment plants with output\r\n(number)' == 0) & ('Output of waste treatment plants\r\n(1000t)' == 0) & ('Wastes to be disposed of\r\n(1000t)' == 0 ) & ('Wastes to be recovered\r\n(1000t)' == 0) &  ('Waste to preparatory processes\r\n(1000t)' == 0) & ('Waste to preparatory processes\r\n(1000t)' == 0) & ('Distribution to other plants0 users or traders\r\n(1000t)' == 0) ]


In [14]:
df['Waste treatment plants\r\n(number)'] = df['Waste treatment plants\r\n(number)'].astype(str)

# Then, filter the DataFrame
df_filtered = df[
    (df['Wastes generated in the own local unit\r\n(1000t)'] == 0) &
    (df['Input to waste treatment plants\r\n(1000t)'] == 0) &
    (df['Waste delivered from the domestic territory\r\n(1000t)'] == 0) &
    (df['Waste delivered from abroad\r\n(1000t)'] == 0) &
    (df['Waste treatment plants\r\n(number)'].str.strip() == '0') &
    (df['Output of waste treatment plants\r\n(1000t)'] == 0) &
    (df['Wastes to be disposed of\r\n(1000t)'] == 0) &
    (df['Wastes to be recovered\r\n(1000t)'] == 0) &
    (df['Waste to preparatory processes\r\n(1000t)'] == 0) &
    (df['Distribution to other plants0 users or traders\r\n(1000t)'] == 0)
]

In [15]:
# pd.set_option('display.max_rows', None)

In [16]:
df_filtered

Year              State                                 Type of plant  \
12    2006  Baden-Württemberg         natural substances produced in mining   
28    2006             Bayern        Wastes at underground extraction sites   
30    2006             Berlin                                     Landfills   
36    2006             Berlin  Mechanical-biological waste treatment plants   
42    2006             Berlin         natural substances produced in mining   
...    ...                ...                                           ...   
3831  2021          Thuringia  Mechanical-biological waste treatment plants   
3835  2021          Thuringia                        Dismantling facilities   
3836  2021          Thuringia                        Other treatment plants   
3837  2021          Thuringia         natural substances produced in mining   
3838  2021          Thuringia        Wastes at underground extraction sites   

     Waste treatment plants\r\n(number)  \
12                                    0   
28                                    0   
30                                    0   
36                                    0   
42                                    0   
...                                 ...   
3831                                  0   
3835                                  0   
3836                                  0   
3837                                  0   
3838                                  0   

      Input to waste treatment plants\r\n(1000t)  \
12                                           0.0   
28                                           0.0   
30                                           0.0   
36                                           0.0   
42                                           0.0   
...                                          ...   
3831                                         0.0   
3835                                         0.0   
3836                                         0.0   
3837                                         0.0   
3838                                         0.0   

      Wastes generated in the own local unit\r\n(1000t)  \
12                                                  0.0   
28                                                  0.0   
30                                                  0.0   
36                                                  0.0   
42                                                  0.0   
...                                                 ...   
3831                                                0.0   
3835                                                0.0   
3836                                                0.0   
3837                                                0.0   
3838                                                0.0   

      Waste delivered from the domestic territory\r\n(1000t)  \
12                                                  0.0        
28                                                  0.0        
30                                                  0.0        
36                                                  0.0        
42                                                  0.0        
...                                                 ...        
3831                                                0.0        
3835                                                0.0        
3836                                                0.0        
3837                                                0.0        
3838                                                0.0        

      Waste delivered from abroad\r\n(1000t)  \
12                                       0.0   
28                                       0.0   
30                                       0.0   
36                                       0.0   
42                                       0.0   
...                                      ...   
3831                                     0.0   
3835                                     0.0   
3836 

In [18]:
1255 / len(df) * 100

32.68229166666667

### Inference -3
There is indeed a strong presence of 0 values. We have rows where entire rows consist of 0  which are 1255 (32% ) of the data soo essentially the least value among the above analysis of the columns. (Waste treatment plants (number) has 1255 Percentage: 32.68 % zeros present) 
Possibilty of imputation - Need to check but seems unlikely as even if I try to find patterns other columns also have 0or na's present
In the site I think all these are due to having  missing values 

### Ideas
We could essentially remove rows which have all the data missing also we could remove columns which have higher percentage of 0 present 
We could impute the columns which have around 30- 50 percent of 0 or missing values on the basis of 3 columns year,state,type and check the distribution of the data